### PR TITLE
Replace usages of `Text` with `Cow<'_, Text>`

### PR DESCRIPTION
**Pull request by DanikVitek**
fixes [#209](https://github.com/valence-rs/valence/issues/209)

### CODE REVIEW

**Review by dyc3 — CHANGES_REQUESTED** [review]
Looks good so far, but there are a bunch of compiler errors if you try to build `valence_new`.
You need to update the examples and some of the methods in `impl Client`.

File: crates/valence_protocol/src/packets/s2c/sound_id.rs [hunk 1]
```diff
@@ -13,6 +13,12 @@ pub enum SoundId<'a> {
     },
 }
 
+impl<'a> From<VarInt> for SoundId<'a> {
+    fn from(id: VarInt) -> Self {
+        SoundId::Reference { id }
+    }
+}
+
```
Comment (dyc3) [inline crates/valence_protocol/src/packets/s2c/sound_id.rs:21]: unrelated

File: crates/packet_inspector/src/main.rs [hunk 2]
```diff
@@ -3,7 +3,7 @@ use std::fmt::Write;
 use std::io::ErrorKind;
 use std::net::SocketAddr;
 use std::sync::Arc;
-use std::{fmt, io};
+use std::io;
```
Comment (dyc3) [inline crates/packet_inspector/src/main.rs:6]: unrelated

File: crates/valence_new/examples/entity_raycast.rs [hunk 3]
```diff
@@ -1,4 +1,3 @@
-use glam::Vec3;
```
Comment (dyc3) [inline crates/valence_new/examples/entity_raycast.rs:1]: unrelated

File: crates/valence_new/src/client.rs [hunk 4]
```diff
@@ -445,12 +446,23 @@ impl Client {
...
+        &'a mut self,
+        url: &'a str,
+        hash: &'a str,
+        forced: bool,
+        prompt_message: Option<impl Into<Cow<'a, Text>>>,
+    ) {
+        self._set_resource_pack(url, hash, forced, prompt_message.map(|t| t.into()))
+    }
+
+    fn _set_resource_pack(
```
Comment (dyc3) [inline crates/valence_new/src/client.rs:460]: Keep as one function.

File: crates/valence_new/src/client.rs [hunk 5]
```diff
@@ -466,15 +478,22 @@ impl Client {
...
-        let subtitle = subtitle.into();
+        self._set_title(title.into(), subtitle.into(), animation.into())
+    }
 
+    fn _set_title(
+        &mut self,
+        title: Cow<Text>,
+        subtitle: Cow<Text>,
+        animation: Option<SetTitleAnimationTimes>,
+    ) {
```
Comment (dyc3) [inline crates/valence_new/src/client.rs:496]: Keep as one function.

File: crates/valence_new/src/client.rs [hunk 6]
```diff
@@ -492,12 +511,18 @@ impl Client {
     ///
     /// The action bar is a small piece of text displayed at the bottom of the
     /// screen, above the hotbar.
-    pub fn set_action_bar(&mut self, text: impl Into<Text>) {
+    #[inline]
+    pub fn set_action_bar<'a>(&'a mut self, text: impl Into<Cow<'a, Text>>) {
+        self._set_action_bar(text.into())
+    }
+
+    fn _set_action_bar(&mut self, text: Cow<Text>) {
```
Comment (dyc3) [inline crates/valence_new/src/client.rs:519]: Keep as one function.

File: crates/valence_new/src/client.rs [hunk 7]
```diff
@@ -492,12 +511,18 @@ impl Client {
...
         self.write_packet(&SetActionBarText {
-            action_bar_text: text.into(),
+            action_bar_text: text,
         });
     }
 
+    #[inline]
```
Comment (dyc3) [inline crates/valence_new/src/client.rs:525]: unrelated change

File: crates/valence_new/src/client.rs [hunk 8]
```diff
@@ -506,12 +531,31 @@ impl Client {
...
+            particle,
+            long_distance,
+            position.into(),
+            offset.into(),
+            max_speed,
+            count,
+        )
+    }
+
+    fn _play_particle(
```
Comment (dyc3) [inline crates/valence_new/src/client.rs:545]: Keep as one function.

File: crates/valence_protocol/src/lib.rs [hunk 9]
```diff
@@ -118,9 +118,23 @@ mod var_long;
...
+        fn encode(&self, _: impl std::io::Write) -> Result<()> {
+            Ok(())
+        }
+    }
+
+    impl<'a, T> Decode<'a> for PhantomData<T> {
+        fn decode(_: &mut &'a [u8]) -> Result<Self> {
+            Ok(PhantomData)
+        }
+    }
```
Comment (dyc3) [inline crates/valence_protocol/src/lib.rs:137]: what is this for?

File: crates/valence_protocol/src/lib.rs [hunk 10]
```diff
@@ -337,52 +351,60 @@ pub trait DecodePacket<'a>: Sized + fmt::Debug {
...
     struct TupleStruct(i32, bool, f64);
 
-    #[derive(Encode, EncodePacket, Decode, DecodePacket)]
+    #[derive(Debug, Encode, EncodePacket, Decode, DecodePacket)]
     #[packet_id = 5]
-    struct StructWithGenerics<'z, T = ()> {
+    struct StructWithGenerics<'z, T = ()>
+    where
+        T: std::fmt::Debug,
+    {
```
Comment (dyc3) [inline crates/valence_protocol/src/lib.rs:379]: unrelated changes

File: crates/valence_new/examples/text.rs [hunk 11]
```diff
@@ -50,10 +50,10 @@ fn init_clients(
 
         client.send_message("Welcome to the text example.".bold());
         client
-            .send_message("The following examples show ways to use the different text components.");
+            .send_message(Text::from("The following examples show ways to use the different text components."));
 
         // Text examples
-        client.send_message("\nText");
+        client.send_message(Text::from("\nText"));
```
Comment (dyc3) [inline crates/valence_new/examples/text.rs:56]: You should make `From` impls so this works without changing all these lines.

**Review by dyc3 — CHANGES_REQUESTED** [review]
There are too many unrelated changes, and that's making this really hard to review. Would you mind removing the changes you made to `packet_inspector` and `valence`? `valence` is broken on this branch and it's going to be removed anyway.

File: crates/valence_new/src/client.rs (hunk 4 above)
Comment (DanikVitek) [reply crates/valence_new/src/client.rs:460]: This would decrease binary size

File: crates/valence_new/src/client.rs (hunk 5 above)
Comment (DanikVitek) [reply crates/valence_new/src/client.rs:496]: As well as this

File: crates/valence_new/examples/text.rs (hunk 11 above)
Comment (DanikVitek) [reply crates/valence_new/examples/text.rs:56]: Ok

File: crates/valence_protocol/src/lib.rs (hunk 9 above)
Comment (DanikVitek) [reply crates/valence_protocol/src/lib.rs:137]: Sorry, unrelated. There was an attempt to make `Proterty<S>` has Cow<'_, str> as a default, but I passed for now

File: crates/valence_new/src/client.rs (hunk 6 above)
Comment (DanikVitek) [reply crates/valence_new/src/client.rs:519]: And this

File: crates/valence_new/src/client.rs (hunk 8 above)
Comment (DanikVitek) [reply crates/valence_new/src/client.rs:545]: And this

File: crates/valence_new/examples/text.rs (hunk 11 above)
Comment (DanikVitek) [reply crates/valence_new/examples/text.rs:56]: It would be not possible to implement From<&str> for Cow<Text>

File: crates/valence_protocol/src/lib.rs (hunk 10 above)
Comment (rj00a) [reply crates/valence_protocol/src/lib.rs:379]: I did but these changes should go in a separate PR to keep things focused.

File: crates/valence_new/src/client.rs (hunk 4 above)
Comment (dyc3) [reply crates/valence_new/src/client.rs:460]: That does not make it any less of an unrelated change. Binary size is not a priority for us, and this PR is about fixing one issue, nothing else.

File: crates/valence_new/examples/text.rs (hunk 11 above)
Comment (dyc3) [reply crates/valence_new/examples/text.rs:56]: Hmm you're right. I still don't want it to be necessary to wrap messages in `Text::from`, that severely degrades dev UX. Maybe you could keep `send_message(text: Into<Text>)` and just make it a `Cow` internally? It might be better to make a wrapper type for `Cow<'a, Text>`. @rj00a thoughts?

Comment (rj00a) [reply crates/valence_new/examples/text.rs:56]: The `Text` API will probably need a thorough redesign at some point. Maybe we could make a dedicated trait like `IntoText` to hide the details between `Text`, `&Text`, `Cow<Text>`, etc. For now, let's just keep `impl Into<Text>` and only change the packets.

**Comment by DanikVitek** [comment]
Fixed merge conflicts

File: crates/valence_new/examples/resource_pack.rs [hunk 12]
```diff
@@ -71,7 +71,7 @@ fn prompt_on_punch(mut clients: Query<&mut Client>, mut events: EventReader<Inte
                 "https://github.com/valence-rs/valence/raw/main/assets/example_pack.zip",
                 "d7c6108849fb190ec2a49f2d38b7f1f897d9ce9f",
                 false,
-                None,
+                Option::<Text>::None,
```
Comment (dyc3) [inline crates/valence_new/examples/resource_pack.rs:74]: Can we revert this back to just `None`?

**Review by dyc3 — APPROVED** [review]
This is looking _much_ better now. Thanks for fixing it up.

File: crates/valence_new/examples/resource_pack.rs (hunk 12 above)
Comment (DanikVitek) [reply crates/valence_new/examples/resource_pack.rs:74]: There is an `Option<impl Into<Text>>` now, so it could be reversed if the function is changed back as well. Should I?

Comment (dyc3) [reply crates/valence_new/examples/resource_pack.rs:74]: Yes, thank you.
